salt/orchestrate/edx/build_ami.sls

```diff
@@ -0,0 +1,176 @@
+  boto_s3_bucket.present:
+    - Bucket: {{ bucket }}-{{ PURPOSE_PREFIX }}-{{ type }}-{{ ENVIRONMENT }}
+    - region: us-east-1
+    - Versioning:
+       Status: "Enabled"
+{% endfor %}
+{% endfor %}
+
+load_pillar_data_on_edx_base_nodes:
+  salt.function:
+    - name: saltutil.refresh_pillar
+    - tgt: 'P@roles:(edx-base|edx-base-worker) and G@environment:{{ ENVIRONMENT }}'
+    - tgt_type: compound
+    - require:
+        - salt: create_edx_worker_baseline_instance_in_{{ ENVIRONMENT }}
+        - salt: create_edx_baseline_instance_in_{{ ENVIRONMENT }}
+
+populate_mine_with_edx_node_data:
+  salt.function:
+    - name: mine.update
+    - tgt: 'P@roles:(edx-base|edx-base-worker) and G@environment:{{ ENVIRONMENT }}'
+    - tgt_type: compound
+    - require:
+        - salt: load_pillar_data_on_edx_base_nodes
+
+{# Reload the pillar data to update values from the salt mine #}
+reload_pillar_data_on_edx_nodes:
+  salt.function:
+    - name: saltutil.refresh_pillar
+    - tgt: 'P@roles:(edx-base|edx-base-worker) and G@environment:{{ ENVIRONMENT }}'
+    - tgt_type: compound
+    - require:
+        - salt: populate_mine_with_edx_node_data
+
+{# Deploy Consul agent first so that the edx deployment can use provided DNS endpoints #}
+deploy_consul_agent_to_edx_nodes:
+  salt.state:
+    - tgt: 'P@roles:(edx-base|edx-base-worker) and G@environment:{{ ENVIRONMENT }}'
+    - tgt_type: compound
+    - sls:
+        - consul
+        - consul.dns_proxy
+
+build_edx_base_nodes:
+  salt.state:
+    - tgt: 'P@roles:(edx-base|edx-base-worker) and G@environment:{{ ENVIRONMENT }}'
+    - tgt_type: compound
+    - highstate: True
+    - require:
+        - salt: deploy_consul_agent_to_edx_nodes
+
+{% set previous_release = salt.sdb.get('sdb://consul/edxapp-release-version')|int %}
+{% set release_number = previous_release + 1 %}
+{% salt.sdb.set('sdb://consul/edxapp-release-version', '{}'.format(release_number) %}
+
+snapshot_edx_app_node:
+  boto_ec2.snapshot_created:
+    - name: edxapp_base_release_{{ release_number }}
+    - ami_name: edxapp_base_release_{{ release_number }}
+    - instance_name: {{ instance_name }}
+    - tags:
+        release_number: {{ release_number }}
+        business_unit: residential
+    - description: MITx application image
+
+snapshot_edx_worker_node:
+  boto_ec2.snapshot_created:
+    - name: edx_worker_base_release_{{ release_number }}
+    - ami_name: edx_worker_base_release_{{ release_number }}
+    - instance_name: {{ worker_instance_name }}
+    - tags:
+        release_number: {{ release_number }}
+        business_unit: residential
+    - description: MITx application image
+
+update_release_version:
+  salt.function:
+    - tgt: 'roles:master'
+    - tgt_type: grain
+    - name: sdb.set
+    - arg:
+        - {{ release_number }}
+    - require:
+        - boto_ec2: snapshot_edx_app_node
+        - boto_ec2: snapshot_edx_worker_node
+
+destroy_edx_base_instance:
```

Comment:
In addition to the two listed requirements, should we add some tests to verify that all's functional before we go ahead and destroy the old instances? Should we also add some notifications on success and failure?